### ISSUE TITLE
Clean up minimiser modules

### ISSUE THREAD

**Issue by lucianopaz** [lucianopaz/slsqp-jax#109]
#107 and #108 were the most rushed and AI assistant reliant features to meet the 2026-08-07 deadline and I'll have to do a cleanup in September

**UPDATE**

I've prepared a write up on what is bad about the current abstraction and how I think it should be fixed. The write up was also AI assisted and is rather verbose, but I think it's helpful for future context because I might forget part of this some time in the near future.

## Summary

The rushed `CommonMinimiser` implementation has two independent defects:

1. It hard-codes one termination formula for algorithms with different mathematical optimality conditions.
2. It commits parts of rejected steps as though they had been accepted.

Fix rejected-step handling first. Then extract termination into:

```text
algorithm-specific measurements
    -> TerminationMetrics
    -> algorithm-specific TerminationFlags
    -> shared classify_termination()
```

The extraction preserves today's residual formulas and subproblem-status lethality. Better residuals and finer failure semantics are follow-up work.

## 1. Current state and what is wrong

The `sqpdax` minimiser sub-package was rushed, and [`CommonMinimiser`](../../slsqp_jax/sqpdax/minimiser/base.py) consequently mixes concerns that differ between concrete algorithms.

### 1.1 `CommonMinimiser` defines termination instead of orchestrating it

`CommonMinimiser.terminate()` currently computes the stationarity test, delegates only feasibility and extra optimality, and folds failure diagnostics over the result:

```python
ctx = self._optimisation_context(problem)
lagrangian = ctx.lagrangian

grad_norm = _inf_norm(lagrangian.x_grad)
scale = jnp.maximum(jnp.abs(lagrangian.value), 1.0)
stationary = grad_norm <= self.rtol * scale
feasible = self._feasibility_error(ctx) <= self.atol
optimal = self._extra_optimality(ctx) <= self.atol
has_min_steps = self.step_count >= self.min_steps
converged = stationary & feasible & optimal & has_min_steps
```

Concrete minimisers can replace the feasibility and extra-optimality values, but not the stationarity measure, norm, or scale. The shared implementation therefore assumes that every constrained algorithm uses `||grad_x L||_inf <= rtol * max(abs(L), 1)`.

That assumption is already questionable for the two live algorithms:

- `ActiveSetLineSearchMinimiser` measures violations of the original NLP constraints and bounds and currently has no extra optimality residual.
- `TrustRegionInteriorPointMinimiser` measures slacked barrier residuals and uses `max(complementarity, mu)` as its extra optimality residual. Its `lagrangian.value` is a barrier value, and the current stationarity test ignores slack variables.

Those mathematical shortcomings motivate the abstraction boundary.

The current shape causes two additional problems:

- Both concrete minimisers contain a `_termination_diagnostics` override with an identical four-line body: promote `solver_state.status`, compare it with `successful`, and append it as a fatal diagnostic.
- The diagnostic fold overwrites `successful`, so a non-success status wins even when the retained iterate satisfies the convergence test. In particular, `ActiveSetQPSolver` can report `max_steps_reached` after the QP working-set budget is exhausted, causing a converged nonlinear iterate to be reported as failed.

### 1.2 Rejected steps are handled incorrectly

`StepResult.accepted` is not read by `CommonMinimiser._execute_step`. The method commits `result.x`, `step_dual`, a secant update, the solver carry, and post-step algorithm dynamics unconditionally:

```python
x_new = result.x
new_secant = self._update_secant(ctx.lagrangian, x_new, step_dual)
advanced = eqx.tree_at(
    lambda m: (m.iterate, m.dual, m.secant, m.solver_state, m.step_count),
    self,
    (x_new, step_dual, new_secant, result.solver_state, self.step_count + 1),
    is_leaf=lambda z: z is None,
)
return advanced._advance_dynamics(ctx, x_new, step_dual)
```

The controllers also disagree about `StepResult` semantics:

- `TrustRegionManager` returns the incoming primal and its merit on rejection, while still returning the reduced trust-region radius in `solver_state`.
- `ArmijoLineSearch` returns its last failed trial and that trial's merit on budget exhaustion, merely setting `accepted=False`.

The latter violates
[`StepResult`](../../slsqp_jax/sqpdax/step_controller/base.py)'s documented contract: `x` should equal the incoming `x0` on rejection, and `merit_val` should be the merit at that returned `x`.

Consequences include:

- A failed Armijo trial can become the next nonlinear iterate.
- A secant pair can be appended for a rejected primal step.
- The interior-point barrier can be updated after trust-region rejection.
- Trust-region QP multipliers can be paired with a retained primal even though the method treats them as one primal-dual iterate.

The solver carry is different: it **must** be committed on rejection because it contains the reduced trust-region radius. Freezing it would repeat the same rejection until the driver exhausts `max_steps`; it would not literally hang, because both drivers have an outer iteration budget.

### 2.1 Correct rejected-step state transitions

Make `StepResult` truthful, preserve rejected primal state, prevent rejected steps from changing secant or accepted-iterate dynamics, and make multiplier selection explicit per algorithm. Continue committing controller/solver carry and counting every outer step attempt.

### 2.2 Reduce `CommonMinimiser.terminate()` to orchestration

The target body is:

```python
def terminate(self, problem):
    ctx = self._optimisation_context(problem)
    metrics = self.termination_metrics(ctx)
    flags = self.termination_flags(metrics)
    return classify_termination(flags)
```

`terminate()` should not define residuals, compare tolerances, or encode result precedence.

This extraction deliberately preserves:

- Current stationarity, feasibility, and extra-optimality formulas.
- Current rule that `optx.RESULTS.promote(solver_state.status) != successful` is fatal.
- Current `rtol`, `atol`, and `min_steps` option surface.

Non-goals:

- No user-swappable termination policy.
- No `TerminationCriterion` ABC or criterion list.
- No kind-registry entry or `termination=` constructor argument.
- No split of `AbstractConstrainedMinimiser` configuration from running state.

Termination remains part of an algorithm's identity. The new public methods are subclass-extension APIs for implementing algorithms, not configuration knobs for callers of `minimise`.

#### Restore the `StepResult` contract

Change `ArmijoLineSearch.step` so that:

- On acceptance, `x` and `merit_val` describe the accepted trial.
- On rejection, `x` is `x0` and `merit_val` is the merit at `x0`, including the `max_steps=0` case.

If rejected-trial diagnostics are needed later, add separately named `trial_x` or `trial_merit_val` fields rather than overloading the committed values.

#### Defensively gate the common commit path

Even with contract-correct controllers, `CommonMinimiser` must enforce its own state-transition invariant because `StepController` is an extension point.

- `iterate`: accept `result.x` only when `result.accepted`; otherwise retain the previous iterate.
- `secant`: append only for an accepted step. Preserve the current static `if self.secant is None` branch; do not use `lax.cond` between `None` and an `LBFGS` instance.
- `solver_state`: always commit `result.solver_state`, including after rejection.
- `step_count`: always increment. It counts attempted outer steps, not only accepted steps, and remains the value used by `min_steps` and `Solution.stats["num_steps"]`.

#### Make multiplier selection algorithm-specific

There is no correct universal rule for the dual on rejection. Add an internal `_select_dual(step_dual, accepted)` hook:

- The safe `CommonMinimiser` default retains `self.dual` on rejection.
- `ActiveSetLineSearchMinimiser` always returns `step_dual`. Its QP multipliers estimate the multipliers at the current linearisation point, so they can improve stationarity at the retained primal.
- `TrustRegionInteriorPointMinimiser` uses the default and retains its previous dual on rejection, keeping the coupled primal-dual iterate intact.

`Dual` has array-only dynamic leaves, so rejection selection can be performed leafwise with `jnp.where`.

#### Run accepted-iterate dynamics only after acceptance

After installing the selected primal/dual/secant and the unconditional solver carry/count, invoke `_advance_dynamics` only in the accepted branch:

```python
return jax.lax.cond(
    result.accepted,
    lambda m: m._advance_dynamics(ctx, committed_x, committed_dual),
    lambda m: m,
    advanced,
)
```

This matters even after fixing `ArmijoLineSearch` because the controller selection cannot prevent `_advance_dynamics` from reducing the barrier.

`_advance_dynamics` keeps its existing signature. Its contract must say that it preserves the minimiser's pytree structure, because both `lax.cond` branches must return the same `Self` treedef. Both current implementations satisfy this after `init`; structural initialization remains the job of
`_init_dynamics`.

### 3.2 Add `slsqp_jax/sqpdax/minimiser/termination.py`

Use one module, not a sub-package. It owns shared data types, a shared nonfinite helper, and result classification. It does not define an algorithm's residuals.

#### `TerminationMetrics`

Add a fixed `equinox.Module`:

```python
class TerminationMetrics(Module):
    stationarity: Scalar
    stationarity_scale: Scalar
    feasibility: Scalar
    extra_optimality: Scalar
    nonfinite: Bool[Array, ""]
    subproblem_result: optx.RESULTS
```

All initial fields are required keyword arguments. A future schema change must update both concrete minimisers and their tests together; do not add speculative defaulted fields.

For this extraction, populate the fields with today's formulas:

- Both algorithms:
  - `stationarity = _inf_norm(lagrangian.x_grad)`.
  - `stationarity_scale = max(abs(lagrangian.value), 1)`.
  - `nonfinite` reproduces today's checks over Lagrangian value and gradient plus every leaf of the iterate and dual.
  - `subproblem_result` is `optx.RESULTS.promote(ctx.solver_state.status)`.
- Active-set line search:
  - `feasibility` is the current maximum NLP equality, inequality, and bound violation.
  - `extra_optimality = 0`.
- Trust-region interior point:
  - `feasibility` is the current maximum over slacked `dual_grad` blocks.
  - `extra_optimality` is today's `max(barrier_update.complementarity(lagrangian), barrier.weight)`.

`extra_optimality` is intentionally a role name. The interior-point value is not raw complementarity, and the active-set zero means “nothing extra is currently measured,” not “complementarity was measured and is zero.”

Do not expose metrics as a dynamic dictionary. A fixed pytree schema is easier to type, test, and carry through JAX transformations.

#### `TerminationFlags`

Add a SQPDAX-local type:

```python
class TerminationFlags(Module):
    converged: Bool[Array, ""]
    nonfinite: Bool[Array, ""]
    fatal: Bool[Array, ""]
    fatal_result: optx.RESULTS
```

`fatal_result` must always be a valid `optx.RESULTS`. When `fatal` is false, set it to `successful`.

#### `classify_termination`

Make precedence executable and testable:

```python
def classify_termination(flags):
    done = flags.nonfinite | flags.converged | flags.fatal
    result = optx.RESULTS.where(
        flags.nonfinite,
        optx.RESULTS.nonfinite,
        optx.RESULTS.where(
            flags.converged,
            optx.RESULTS.successful,
            optx.RESULTS.where(
                flags.fatal,
                flags.fatal_result,
                optx.RESULTS.successful,
            ),
        ),
    )
    return done, result
```

Precedence is:

1. Nonfinite state.
2. Valid convergence.
3. Fatal algorithm/subproblem result.
4. `successful` while still running.

Convergence therefore beats a stale failure left by the preceding subproblem attempt, while nonfinite always wins. The driver continues to map `done=False` at outer budget exhaustion to `nonlinear_max_steps_reached`.

#### `termination_metrics(ctx) -> TerminationMetrics`

Make this abstract on `CommonMinimiser` and implement it on each concrete minimiser. It replaces `_feasibility_error`, `_extra_optimality`, and the concrete `_termination_diagnostics` methods by collecting all relevant measurements in one place.

`OptimisationContext` is already public, exported, and constructible. The method's numpy-style docstring should nevertheless direct subclass authors to `_optimisation_context` as the normal way to obtain a populated context.

#### `termination_flags(metrics) -> TerminationFlags`

Provide a concrete, overridable implementation on `CommonMinimiser`:

```python
stationary = metrics.stationarity <= self.rtol * metrics.stationarity_scale
feasible = metrics.feasibility <= self.atol
optimal = metrics.extra_optimality <= self.atol
has_min_steps = self.step_count >= self.min_steps

converged = stationary & feasible & optimal & has_min_steps
fatal = metrics.subproblem_result != optx.RESULTS.successful
fatal_result = optx.RESULTS.where(
    fatal,
    metrics.subproblem_result,
    optx.RESULTS.successful,
)

return TerminationFlags(
    converged=converged,
    nonfinite=metrics.nonfinite,
    fatal=fatal,
    fatal_result=fatal_result,
)
```

The explicit `fatal_result` selection matters for future recoverable statuses: `fatal` may eventually be false while `subproblem_result` remains non-successful.

`termination_flags` remains a bound method so future algorithm-specific history can be read from `self` without changing this public signature.

Both public methods need numpy-style docstrings.

### 3.4 Simplify `CommonMinimiser.terminate()`

After the extraction, `terminate()` performs only:

1. Context construction.
2. Metric extraction.
3. Flag construction.
4. Shared classification.

Delete `_feasibility_error`, `_extra_optimality`, and `_termination_diagnostics`. Move `_nonfinite_detected` to `termination.py` as a shared free helper rather than deleting the functionality.

Keep `rtol`, `atol`, and `min_steps` where they are. This leaves option validation and `as_optimistix_minimiser` unchanged.

### 4.1 `TerminationHistory`

`TerminationHistory` does not exist today. `OptimisationContext`, `TerminationMetrics`, `TerminationFlags`, and `classify_termination` describe one current iterate:

- The context provides evaluated algorithm state.
- Metrics normalize the algorithm-specific measurements into named roles.
- Flags state which current-iterate outcomes fired.
- Classification gives those flags a shared result precedence.

None remembers prior iterations. History becomes necessary only for conditions that depend on a sequence:

- Merit stagnation over an N-step window.
- K consecutive QP or line-search failures.
- Divergence relative to the best merit seen.
- Consecutive converged zero steps.
- Restoration entry count, cooldown, or violation stall.

At that point, `TerminationHistory` would be a small, fixed-shape pytree stored on the minimiser beside `iterate`, `dual`, `secant`, `solver_state`, and `step_count`. It would contain only fields consumed by implemented rules.

Its ownership and update discipline are:

- `_execute_step` writes it exactly once per outer step attempt.
- `termination_flags(metrics)` reads it from `self.termination_history`.
- `terminate()` and `classify_termination()` never mutate it.

This is required because the driver calls `terminate()` in the loop condition and again after the loop. Updating history inside `terminate()` would double-count one solver state.

Future merit-stagnation logic, for example, must use:

```python
improved = result.accepted & (result.merit_val < history.best_merit)
```

A rejected trial must not reset accepted-iterate progress. If diagnostics need the rejected trial's merit, `StepResult` should carry it under a distinct name. Likewise, rollback must define the full state snapshot it restores, rather than storing only a best scalar merit.

### 4.2 Do not reuse legacy termination classes

Although [`slsqp/termination.py`](../../slsqp_jax/slsqp/termination.py) has similarly named types, SQPDAX should remain independent:

- `sqpdax` currently has no imports from legacy top-level solver modules.
- Legacy classification returns `slsqp_jax.RESULTS`, a subclass of `optx.RESULTS`. Cross-class equality raises under Equinox enumeration semantics, so mixing the two types is unsafe.
- Legacy `coarse_outcome` owns maximum-iteration termination, while SQPDAX's driver already owns the outer budget.
- The legacy `TerminationFlags` has nine fields, several of which have no SQPDAX source yet.

Define local SQPDAX types using `optx.RESULTS`. Structural similarity may be noted in docstrings without creating a package dependency.

### 4.3 Keep QP budget semantics out of this refactor

During extraction, retain the exact current fatal predicate:

```python
metrics.subproblem_result != optx.RESULTS.successful
```

Do not additionally inspect `solver_state.success`; it is currently false in some initial or exhausted states whose `status` is successful.

The active-set solver also needs separate follow-up work before `max_steps_reached` can safely become recoverable:

- Its carried `n_iter` currently also controls the next solve's local working set budget. Once the cumulative count reaches `qp_max_iter`, later calls can perform no new working-set iterations.
- Working-set exhaustion and inner-solver iteration exhaustion are not represented distinctly enough for the outer minimiser to choose different recovery policies.

Changing lethality now could turn a visible failure into repeated zero-work outer attempts. First separate per-solve budget from cumulative diagnostics and give incomplete outcomes distinct status/state.

### 4.4 Explicit follow-up numerical work

The extraction creates the seam for, but does not implement:

- FilterSQP `mu_max` scaling for active-set stationarity.
- A complete interior-point residual and a justified scale, coordinated with
  `BarrierUpdate.optimality_residual`.
- Active-set dual-feasibility and complementarity checks.
- Recoverable QP/line-search failure patience.
- Guarded zero-step QP-KKT convergence.
- Stagnation, divergence rollback, and restoration history.

These should arrive as separate numerical changes with dedicated tests, not as
incidental behavior changes in the architectural refactor.

### 4.5 Observable behavior changes

This proposal intentionally makes exactly four behavior changes:

1. Rejected Armijo results satisfy the `StepResult` contract.
2. Rejected steps retain primal/secant state and do not advance barrier dynamics, while still committing solver carry and attempt count.
3. Trust-region interior-point rejection retains the old dual; active-set rejection continues to install the fresh QP multiplier estimate.
4. Valid convergence beats a stale non-success subproblem status, except that nonfinite state has highest priority.

Everything else in the termination extraction is a move of current formulas.

### 4.6 Tests and documentation

Add focused tests for:

- Armijo rejection, including `max_steps=0`, returns `x0` and its finite merit.
- Common rejection retains iterate and secant but commits solver carry and increments `step_count`.
- Active-set rejection installs the new QP multipliers.
- Trust-region rejection retains the old dual and barrier while committing the reduced radius.
- Accepted steps update all accepted-state fields.
- Every overlap in the classification truth table, especially `(nonfinite, converged, fatal)` and `(converged, fatal)`.
- `max_steps_reached` remains fatal when not converged and is overridden by valid convergence.
- Eager and JIT execution through both the owned driver and Optimistix adapter.

Update affected docstrings:

- `StepResult` and `ArmijoLineSearch` rejection semantics.
- `step_count` as attempted outer steps.
- `_advance_dynamics` as an accepted-step, structure-preserving hook.
- `OptimisationContext` and the new public subclass-extension methods.
- Any `TrustRegionManager` wording that says rejected steps need no outer-loop special handling.
